src/components/button/index.tsx

```diff
@@ -4,37 +4,71 @@ import {
   ButtonProps as ChakraButtonProps,
 } from '@chakra-ui/react';
 
-type WithOnClick<T extends 'submit' | 'button'> = T extends 'submit'
-  ? Pick<ChakraButtonProps, 'onClick'>
-  : { onClick: Exclude<ChakraButtonProps['onClick'], undefined> };
-
 type SharedProps = {
   variant?: 'primary' | 'secondary';
   size?: 'md' | 'lg';
-} & Pick<ChakraButtonProps, 'isDisabled' | 'children'>;
+} & Pick<ChakraButtonProps, 'children'>;
+
+type MandatoryOnClick = {
+  as?: 'button';
+  type: 'submit';
+  onClick?: () => void;
+  href?: never;
+  isExternal?: never;
+  rel?: never;
+} & SharedProps &
+  Pick<ChakraButtonProps, 'isDisabled'>;
 
-type ButtonTypeProps = SharedProps & {
+type OptionalOnClick = {
+  as?: 'button';
   type?: 'button';
-} & WithOnClick<'button'>;
+  onClick?: () => void;
+  href?: never;
+  isExternal?: never;
+  rel?: never;
+} & SharedProps &
+  Pick<ChakraButtonProps, 'isDisabled'>;
 
-type SubmitTypeProps = SharedProps & {
-  type: 'submit';
-} & WithOnClick<'submit'>;
-
-type DefaultTypeProps = SharedProps & WithOnClick<'button'>;
-
-export type ButtonProps = ButtonTypeProps | SubmitTypeProps | DefaultTypeProps;
-
-const Button: FC<ButtonProps> = ({
-  variant = 'primary',
-  size = 'lg',
-  isDisabled = false,
-  children,
-  ...rest
-}) => (
-  <ChakraButton isDisabled={isDisabled} variant={variant} size={size} {...rest}>
-    {children}
-  </ChakraButton>
-);
+export type ButtonProps = MandatoryOnClick | OptionalOnClick;
+
+type LinkProps = {
+  as: 'a';
+  href: string;
+  isExternal?: boolean;
+  rel?: string;
+  isDisabled?: false;
+} & SharedProps;
+
+export type Props = LinkProps | ButtonProps;
+
+const Button: FC<Props> = ({ children, ...props }) => {
```

Comment:
I would directly destructure them 
```
const Button: FC<Props> = ({
  children,
  variant = 'primary',
  size = 'lg',
  isDisabled = false,
  as = 'button',
  ...props
})
```

Reply:
Thanks for exaplainging.

Reply:
Yes, you can destructure. But as soon as the `as` prop is restructured you don't have a prop that differentiates `Link` and `Button` props. To be able to access props that are only available in one part of the union you need a way to show the type system that the object you're accessing it on is of the correct type (in this case `LinkProps`, but the only common part of `Link` and `Button` props is `as` that's no longer there on the `rest`).

[`in` type guard](https://learntypescript.dev/07/l5-in-type-guard) is an alternative to using `as` prop that you could try though. But it's up to you.

Reply:
Distructured them directly avoid the issues like react does not recognize some properties on a DOM element.